File: ios/Plugin/Security/SecurityManager.swift
import Foundation
import Security
import CommonCrypto
import CryptoKit

class SecurityManager {
    private var config: [String: Any]?
    private let keychain = KeychainWrapper()
    
    func configure(_ config: [String: Any]) throws {
        self.config = config
        
        // Validate security configuration
        let enforceHttps = config["enforceHttps"] as? Bool ?? true
        if !enforceHttps {
            print("⚠️ SecurityManager: HTTPS enforcement is disabled. This is not recommended for production.")
        }
    }
    
    func getSecurityInfo() -> [String: Any] {
        return [
            "enforceHttps": config?["enforceHttps"] as? Bool ?? true,
            "certificatePinning": [
                "enabled": (config?["certificatePinning"] as? [String: Any])?["enabled"] as? Bool ?? false,
                "certificates": (config?["certificatePinning"] as? [String: Any])?["certificates"] as? [String] ?? []
            ],
            "validateInputs": config?["validateInputs"] as? Bool ?? true,
            "secureStorage": config?["secureStorage"] as? Bool ?? true
        ]
    }
    
    func validateUrl(_ url: String) -> Bool {
        if !url.hasPrefix("https://") && isHttpsEnforced() {
            return false
        }
        
        // Check against allowed hosts if configured
        if let allowedHosts = getAllowedHosts(), !allowedHosts.isEmpty {
            guard let urlComponents = URLComponents(string: url),
                  let host = urlComponents.host else {
                return false
            }
            return allowedHosts.contains(host)
        }
        
        return true
    }
    
    func verifySignature(data: Data, signature: String, publicKeyString: String) -> Bool {
        guard let publicKeyData = Data(base64Encoded: publicKeyString),
              let signatureData = Data(base64Encoded: signature) else {
            return false
        }
        
        do {
            // Create SecKey from public key data
            let attributes: [String: Any] = [
                kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
                kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
                kSecAttrKeySizeInBits as String: 2048
            ]
            
            guard let secKey = SecKeyCreateWithData(publicKeyData as CFData, attributes as CFDictionary, nil) else {
                return false
            }
            
            // Verify signature
            let algorithm = SecKeyAlgorithm.rsaSignatureMessagePKCS1v15SHA256
            return SecKeyVerifySignature(secKey, algorithm, data as CFData, signatureData as CFData, nil)
        } catch {
            print("Signature verification failed: \(error)")
            return false
        }
    }
    
    func calculateChecksum(for data: Data, algorithm: String = "SHA-256") -> String {
        let digest: Data
        
        switch algorithm {
        case "SHA-256":
            digest = SHA256.hash(data: data).data
        case "SHA-512":
            digest = SHA512.hash(data: data).data
        default:
            // Fallback to SHA-256
            digest = SHA256.hash(data: data).data
        }
        
        return digest.map { String(format: "%02x", $0) }.joined()
    }
    
    func saveSecureData(key: String, value: String) {
        if isSecureStorageEnabled() {
            keychain.set(value, forKey: key)
        } else {
            // Fallback to UserDefaults (not recommended)
            UserDefaults.standard.set(value, forKey: key)
        }
    }
    
    func getSecureData(key: String) -> String? {
        if isSecureStorageEnabled() {
            return keychain.string(forKey: key)
        } else {
            return UserDefaults.standard.string(forKey: key)
        }
    }
    
    func validatePath(_ path: String) -> Bool {
        // Prevent directory traversal attacks
        if path.contains("..") || path.contains("//") {
            return false
        }
        
        // Ensure path is within app's sandbox
        let url = URL(fileURLWithPath: path)
        let standardizedPath = url.standardizedFileURL.path
        
        // Check if path is within allowed directories
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.path
        let cachesPath = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!.path
        let tempPath = NSTemporaryDirectory()
        
        return standardizedPath.hasPrefix(documentsPath) ||
               standardizedPath.hasPrefix(cachesPath) ||
               standardizedPath.hasPrefix(tempPath)
    }
    
    func sanitizeInput(_ input: String) -> String {
        // Remove potentially dangerous characters
        let allowedCharacters = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._-"))
        return input.components(separatedBy: allowedCharacters.inverted).joined()
    }
    
    func isHttpsEnforced() -> Bool {
        return config?["enforceHttps"] as? Bool ?? true
    }
    
    func isSecureStorageEnabled() -> Bool {
        return config?["secureStorage"] as? Bool ?? true
    }
    
    func isInputValidationEnabled() -> Bool {
        return config?["validateInputs"] as? Bool ?? true
    }
    
    func logSecurityEvent(_ event: String, details: String? = nil) {
        guard config?["logSecurityEvents"] as? Bool == true else { return }
        
        let message = "Security Event: \(event) \(details ?? "")"
        print("🔒 \(message)")
        
        // In production, you might want to send these to a security monitoring service
    }
    
    private func getAllowedHosts() -> [String]? {
        var hosts: [String] = []
        
        // Add hosts from live update config
        if let liveUpdateConfig = config?["liveUpdate"] as? [String: Any],
           let allowedHosts = liveUpdateConfig["allowedHosts"] as? [String] {
            hosts.append(contentsOf: allowedHosts)
        }
        
        // Add host from server URL
        if let serverUrl = config?["serverUrl"] as? String,
           let urlComponents = URLComponents(string: serverUrl),
           let host = urlComponents.host {
            hosts.append(host)
        }
        
        return hosts.isEmpty ? nil : Array(Set(hosts))
    }
}

// MARK: - Keychain Wrapper

private class KeychainWrapper {
    private let serviceName = "com.capacitor.nativeupdate"
    
    func set(_ value: String, forKey key: String) {
        guard let data = value.data(using: .utf8) else { return }
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: serviceName,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]
        
        // Delete any existing item
        SecItemDelete(query as CFDictionary)
        
        // Add new item
        SecItemAdd(query as CFDictionary, nil)
    }
    
    func string(forKey key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: serviceName,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var dataTypeRef: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &dataTypeRef)
        
        guard status == errSecSuccess,
              let data = dataTypeRef as? Data,
              let string = String(data: data, encoding: .utf8) else {
            return nil
        }
        
        return string
    }
    
    func remove(forKey key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: serviceName,
            kSecAttrAccount as String: key
        ]
        
        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - CryptoKit Extensions

extension Digest {
    var data: Data {
        Data(self)
    }
}

File: android/src/main/java/com/capacitor/nativeupdate/SecurityManager.kt
package com.capacitor.nativeupdate

import android.content.Context
import android.util.Base64
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.getcapacitor.JSObject
import okhttp3.CertificatePinner
import java.io.File
import java.security.*
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher

class SecurityManager(private val context: Context) {
    private var config: JSObject? = null
    private val masterKey: MasterKey
    private val securePrefs: android.content.SharedPreferences
    
    init {
        // Initialize master key for encryption
        masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        
        // Initialize encrypted shared preferences
        securePrefs = EncryptedSharedPreferences.create(
            context,
            "native_update_secure",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
    
    fun configure(config: JSObject) {
        this.config = config
        
        // Validate security configuration
        val enforceHttps = config.getBool("enforceHttps") ?: true
        if (!enforceHttps) {
            // Log warning about disabled HTTPS enforcement
            android.util.Log.w("SecurityManager", "HTTPS enforcement is disabled. This is not recommended for production.")
        }
    }
    
    fun getSecurityInfo(): JSObject {
        val result = JSObject()
        
        result.put("enforceHttps", config?.getBool("enforceHttps") ?: true)
        
        val certificatePinning = JSObject()
        val pinningConfig = config?.getJSObject("certificatePinning")
        certificatePinning.put("enabled", pinningConfig?.getBool("enabled") ?: false)
        
        // Convert pins to proper format
        val pins = mutableListOf<JSObject>()
        val pinsArray = pinningConfig?.getJSONArray("pins")
        if (pinsArray != null) {
            for (i in 0 until pinsArray.length()) {
                val pin = pinsArray.getJSONObject(i)
                val pinObject = JSObject()
                pinObject.put("hostname", pin.getString("hostname"))
                pinObject.put("sha256", pin.getJSONArray("sha256"))
                pins.add(pinObject)
            }
        }
        certificatePinning.put("pins", pins.toTypedArray())
        result.put("certificatePinning", certificatePinning)
        
        result.put("validateInputs", config?.getBool("validateInputs") ?: true)
        result.put("secureStorage", config?.getBool("secureStorage") ?: true)
        
        return result
    }
    
    fun validateUrl(url: String): Boolean {
        if (!url.startsWith("https://") && isHttpsEnforced()) {
            return false
        }
        
        // Check against allowed hosts if configured
        val allowedHosts = getAllowedHosts()
        if (allowedHosts.isNotEmpty()) {
            val uri = android.net.Uri.parse(url)
            val host = uri.host ?: return false
            return allowedHosts.contains(host)
        }
        
        return true
    }
    
    fun getCertificatePinner(): CertificatePinner? {
        val pinningConfig = config?.getJSObject("certificatePinning") ?: return null
        
        if (pinningConfig.getBool("enabled") != true) {
            return null
        }
        
        val pins = pinningConfig.getJSONArray("pins") ?: return null
        if (pins.length() == 0) {
            return null
        }
        
        val builder = CertificatePinner.Builder()
        
        // Add certificate pins for each host
        for (i in 0 until pins.length()) {
            val pin = pins.getJSONObject(i)
            val hostname = pin.getString("hostname") ?: continue
            val sha256Pins = pin.getJSONArray("sha256") ?: continue
            
            for (j in 0 until sha256Pins.length()) {
                val sha256Pin = sha256Pins.getString(j)
                builder.add(hostname, sha256Pin)
            }
        }
        
        return builder.build()
    }
    
    fun verifySignature(data: ByteArray, signature: String, publicKeyString: String): Boolean {
        return try {
            // Handle PEM format or base64 encoded public key
            val publicKeyBase64 = if (publicKeyString.contains("-----BEGIN PUBLIC KEY-----")) {
                publicKeyString
                    .replace("-----BEGIN PUBLIC KEY-----", "")
                    .replace("-----END PUBLIC KEY-----", "")
                    .replace("\n", "")
                    .replace("\r", "")
                    .trim()
            } else {
                publicKeyString
            }
            
            val publicKeyBytes = Base64.decode(publicKeyBase64, Base64.DEFAULT)
            val keySpec = X509EncodedKeySpec(publicKeyBytes)
            val keyFactory = KeyFactory.getInstance("RSA")
            val publicKey = keyFactory.generatePublic(keySpec)
            
            // Use RSA-PSS to match web implementation
            val sig = Signature.getInstance("SHA256withRSA/PSS")
            sig.initVerify(publicKey)
            sig.update(data)
            
            val signatureBytes = Base64.decode(signature, Base64.DEFAULT)
            sig.verify(signatureBytes)
        } catch (e: Exception) {
            android.util.Log.e("SecurityManager", "Signature verification failed", e)
            false
        }
    }
    
    fun calculateChecksum(file: File, algorithm: String = "SHA-256"): String {
        val digest = MessageDigest.getInstance(algorithm)
        file.inputStream().use { input ->
            val buffer = ByteArray(8192)
            var read: Int
            while (input.read(buffer).also { read = it } != -1) {
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }
    
    fun saveSecureData(key: String, value: String) {
        if (isSecureStorageEnabled()) {
            securePrefs.edit().putString(key, value).apply()
        } else {
            // Fallback to regular preferences (not recommended)
            val prefs = context.getSharedPreferences("native_update", Context.MODE_PRIVATE)
            prefs.edit().putString(key, value).apply()
        }
    }
    
    fun getSecureData(key: String): String? {
        return if (isSecureStorageEnabled()) {
            securePrefs.getString(key, null)
        } else {
            val prefs = context.getSharedPreferences("native_update", Context.MODE_PRIVATE)
            prefs.getString(key, null)
        }
    }
    
    fun validatePath(path: String): Boolean {
        // Prevent directory traversal attacks
        if (path.contains("..") || path.contains("//")) {
            return false
        }
        
        // Ensure path is within app's private directory
        val file = File(path)
        val canonicalPath = file.canonicalPath
        val appDir = context.filesDir.canonicalPath
        
        return canonicalPath.startsWith(appDir)
    }
    
    fun sanitizeInput(input: String): String {
        // Remove potentially dangerous characters
        return input.replace(Regex("[^a-zA-Z0-9._-]"), "")
    }
    
    fun isHttpsEnforced(): Boolean {
        return config?.getBool("enforceHttps") ?: true
    }
    
    fun isSecureStorageEnabled(): Boolean {
        return config?.getBool("secureStorage") ?: true
    }
    
    fun isInputValidationEnabled(): Boolean {
        return config?.getBool("validateInputs") ?: true
    }
    
    private fun getAllowedHosts(): List<String> {
        val hosts = mutableListOf<String>()
        
        // Add hosts from certificate pinning config
        val pinningConfig = config?.getJSObject("certificatePinning")
        if (pinningConfig != null) {
            // In a real implementation, we'd extract hosts from the configuration
            // For now, we'll use the server URL
            val serverUrl = config?.getString("serverUrl")
            if (serverUrl != null) {
                val uri = android.net.Uri.parse(serverUrl)
                uri.host?.let { hosts.add(it) }
            }
        }
        
        // Add allowed hosts from live update config
        val liveUpdateConfig = config?.getJSObject("liveUpdate")
        val allowedHosts = liveUpdateConfig?.getJSONArray("allowedHosts")
        if (allowedHosts != null) {
            for (i in 0 until allowedHosts.length()) {
                hosts.add(allowedHosts.getString(i))
            }
        }
        
        return hosts.distinct()
    }
    
    fun logSecurityEvent(event: String, details: String? = null) {
        if (config?.getBool("logSecurityEvents") == true) {
            android.util.Log.i("SecurityManager", "Security Event: $event ${details ?: ""}")
            
            // In production, you might want to send these to a security monitoring service
        }
    }
}